File: acm-center/poj/2915/2915.py
#coding:utf8
from copy import deepcopy as clone
from spider import *

class group:
	def __init__(self, color='A', cnt=0):
		self.color=color
		self.cnt=cnt
	def __repr__(self):
		return "%s%d" % (self.color, self.cnt)
	def __len__(self):
		return self.cnt
		
class grouplist:
	def __init__(self, gs=None):
		if gs is None:
			gs=[]
		assert type(gs) is list
		self.groups=gs
		
	def __add__(self, y):
		if isinstance(y, group):
			return grouplist(self.groups+[y])
		elif isinstance(y, grouplist):
			return grouplist(self.groups+y.groups)
		elif type(y) is list:
			return grouplist(self.groups+y)
		else:
			raise Exception('can not add')
		
	def __iadd__(self, y):
		if isinstance(y, group):
			self.groups.append(y)
		elif isinstance(y, grouplist):
			self.groups.extend(y.groups)
		elif type(y) is list:
			self.groups.extend(y)
		else:
			raise Exception('can not iadd')
		return self
		
	def __repr__(self):
		return '_'.join([str(i) for i in self.groups])
		
	def __len__(self):
		return len(self.groups)
		
	def __lt__(self, r):
		if len(self)!=len(r):
			return len(self)<len(r)
		else:
			return str(self)<str(r)
	
	def __le__(self, r):
		if len(self)!=len(r):
			return len(self)<len(r)
		else:
			return str(self)<=str(r)
			
	def __gt__(self, r):
		if len(self)!=len(r):
			return len(self)>len(r)
		else:
			return str(self)>str(r)
	
	def __ge__(self, r):
		if len(self)!=len(r):
			return len(self)>len(r)
		else:
			return str(self)>=str(r)
			
	def __eq__(self, r):
		if len(self)!=len(r):
			return False
		else:
			return str(self)==str(r)
		
	def __ne__(self, r):
		if len(self)!=len(r):
			return True
		else:
			return str(self)!=str(r)
			
	def __cmp__(self, r):
		return self<r
		
	def __hash__(self):
		return hash(str(self))
		
	def __getitem__(self, i):
		return self.groups[i]
		
	def __getslice__(self, i, j):
		assert isinstance(i, int)
		assert isinstance(j, int)
		return grouplist(self.groups[i:j])
		
	def insert(self, i, g):
		assert isinstance(i, int)
		assert isinstance(g, group)
		return self.groups.insert(i, g)
		
	def append(self, g):
		assert isinstance(g, group)
		return self.groups.append(g)
		
def B(g):
	assert g.cnt>=1
	# print 'B add', 2 if g.cnt==1 else 1, 'for', g
	return 2 if g.cnt==1 else 1
	
def B2(g):
	assert g.cnt>=2
	# print 'B2 add', 1 if g.cnt==2 else 0, 'for', g
	return 1 if g.cnt==2 else 0
	
mf={}
mg={}
sf={}
sg={}

def F(gs, layer=0):
	# print 'F layer=', layer
	if gs in mf:
		# print 'find1:', gs, '=', mf[gs]
		return mf[gs]
	l=len(gs)
	r=0
	if l==1:
		C=gs[0]
		r=B(C)
	elif l==2:
		L=gs[0]
		R=gs[1]
		r=B(L)+B(R)
	else:
		L=gs[0]
		R=gs[-1]
		if L.color==R.color:
			C=gs[1:-1]
			r=B2(group(L.color, (L.cnt+R.cnt)))+F(C)
		else:
			LC=gs[:-1]
			CR=gs[1:]
			x=B(L)+F(CR, layer+1)
			y=F(LC, layer+1)+B(R)
			r=min(x, y)
	if not gs in mf:
		mf[gs]=r
	# print 'result1:', gs, '=', r
	return r

def G(gs, layer=0):
	# print gs
	# print 'G layer=', layer
	if gs in mg:
		# print 'find2:', gs, '=', mg[gs]
		return mg[gs]
	l=len(gs)
	r=0
	if l==0:
		r=0
		sg[gs]='all explode'
	elif l<=3:
		r=F(gs)
		sg[gs]='F(%s)=%d' % (gs, r)
	else:
		r=F(gs)
		flag=['F(%s)=%d' % (gs, r)]
		n=l
		for radius in range(1, n-1+1):	
			for p in range(1, n-radius+1+1):
				# print 'n=', n, 'p=', p
				C=gs[p-1:p-1+radius]
				# print 'C=', C
				L=gs[:p-1]
				# print 'L=', L
				R=gs[p-1+radius:]
				# print 'R=', R
				
				oL=clone(L)
				oR=clone(R)
				#merge L, R
				while len(L)>0 and len(R)>0 and L[-1].color==R[0].color and L[-1].cnt+R[0].cnt>=3:
					# print 'enter while:\t', L, R
					del L.groups[-1]
					del R.groups[0]
					# print 'after del:\t', L, R
				if len(L)>0 and len(R)>0 and L[-1].color==R[0].color and L[-1].cnt+R[0].cnt<3:
					# print 'enter if:\t', L, R
					# L.groups[-1].cnt+=R.groups[0].cnt
					# del R.groups[0]
					g=R.groups[0]
					tmp=group(g.color, g.cnt)
					tmp.cnt+=L.groups[-1].cnt
					del L.groups[-1]
					del R.groups[0]
					L.groups.append(tmp)
					# print 'after del:\t', L, R
				U=L+R
				# print 'U=', U
				x=F(C)+G(U, layer+1)
				if x<r:
					r=x
					flag=['L(%s)+F(%s)+R(%s) ~ L(%s)+R(%s) => F(%s)+G(%s)=%d' % (oL, C, oR, L, R, C, U, r)]
				elif x==r:
					flag.append('L(%s)+F(%s)+R(%s) ~ L(%s)+R(%s) => F(%s)+G(%s)=%d' % (oL, C, oR, L, R, C, U, r))
		sg[gs]=flag
	if not gs in mg:
		mg[gs]=r
	# print 'result2:', gs, '=', r
	return r
	
def solve(s):
	# print s
	gl=[]
	l=len(s)
	g=group()
	for i, c in enumerate(s):
		if g.cnt==0:
			g.color=c
			g.cnt=1
		if i+1<l:
			if s[i+1]==c:
				g.cnt+=1
			else:
				gl.append(g)
				g=group()
		else:
			gl.append(g)
	gs=grouplist(gl)
	time_seed(0)
	print gs, G(gs)
	print time_gap(0, 'time')
	# print gs
	# print sg[gs]
	
def main():
	for i, s in enumerate(open('data.txt').read().split()):
		if i%2==1: solve(s) 
	
def test_group_list():
	gs=grouplist()
	gs1=grouplist()
	gs2=grouplist()
	gs.groups.extend([group('A', 3), group('A', 1), group('A', 1)])
	# gs+=[group('A', 1), group('A', 1), group('B', 1)]
	gs1.groups.extend([group('A', 2), group('A', 1)])
	gs2.groups.extend([group('A', 2), group('A', 1)])
	l=[gs, gs1, gs2]
	print l
	l.sort()
	print l
	
	m={}
	m[gs]=1
	m[gs1]=2
	gs2+=group('X', 100)
	print gs2
	m[gs2]=3
	print m
	print m.keys()[0][0].cnt
	print gs
	print gs[0]
	print gs[0].color
	print set([gs, gs1, gs2])
	print gs+gs1
	print gs
	print gs1
	
def test_group_list2():
	a=grouplist()
	b=grouplist()
	a.groups.extend([group('A', 3), group('A', 1), group('A', 1)])
	# a+=[group('A', 1), group('A', 1), group('B', 1)]
	b.groups.extend([group('A', 2), group('A', 1)])
	l=[a, b]
	print l
	l.sort()
	print l
	print a
	print b
	print a+b
	print b+a
	print a
	print b
	a+=b
	print a
	print b
	
if __name__=='__main__':
	main()
	# test_group_list()
	# test_group_list2()
	# print
	# for i in mf:
		# print mf[i], i
	# print 
	# for i in mg:
		# print mg[i], i
	# print 
	# for i in sg:
		# print i, ':\n\t', sg[i]

File: acm-center/poj/2915/2915.cpp
/**
 * File: 2915.cpp
 * Description: 
 *
 * Copyright (c) 2011 Jadesoul (Home Page: http://jadesoul.org)
 * Licensed under the GNU General Public License:
 * http://www.gnu.org/licenses/gpl.html
 * 
 * Date: 2011-08-13 15:06:20.571000
 * Written In: Peking University, beijing, China
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
using namespace std;

#define debug
#ifdef debug
	#include <cassert>
	#include <jadesoul>
	ifstream in("data.txt");
	#define print(...) cout<<#__VA_ARGS__<<"="<<__VA_ARGS__<<endl;
	// #define print(...)
	#define dbgcode(...) __VA_ARGS__
	#define dbglog(...) log(__VA_ARGS__)
#else
	istream& in=cin;
	#define print(...)
	#define dbgcode(...)
	#define dbglog(...)
#endif

class group {
public:
	char color;
	int cnt;
	group(char color='A', int cnt=0) {
		this->color=color;
		this->cnt=cnt;
	}
	string tostr() {
		ostringstream oss;
		output(oss);
		return oss.str();
	}
	size_t len() const {
		return cnt;
	}
	friend ostream& operator<<(ostream& os, group& g) {
		return g.output(os);
	}
private:
	ostream& output(ostream& os) {
		return os<<color<<cnt;
	}
};

typedef vector<group> groupvector;

inline groupvector operator+(const groupvector& l, const groupvector& r) {
	groupvector tmp(l);
	tmp.insert(tmp.end(), r.begin(), r.end());
	return tmp;
}

inline void extend(groupvector& me, const groupvector& r) {
	if (r.size()==0) return;
	me.insert(me.end(), r.begin(), r.end());
}

class grouplist {
private:
	groupvector groups;
	string hash;
	void update_hash() {
		this->hash=tostr();
	}
public:
	
	grouplist() {
		groups=groupvector();
		update_hash();
	}
	grouplist(const groupvector& gs) {
		groups=gs;
		update_hash();
	}
	const grouplist operator+(const group& g) {
		groupvector tmp(1, g);
		return grouplist(this->groups+tmp);
	}
	const grouplist operator+(const grouplist& r) {
		return grouplist(this->groups+r.groups);
	}
	const grouplist operator+(const groupvector& v) {
		return grouplist(this->groups+v);
	}
	grouplist& operator+=(const group& g) {
		this->groups.push_back(g);
		update_hash();
		return *this;
	}
	grouplist& operator+=(const grouplist& r) {
		extend(this->groups, r.groups);
		update_hash();
		return *this;
	}
	grouplist& operator+=(const groupvector& v) {
		extend(this->groups, v);
		update_hash();
		return *this;
	}
	string tostr() {
		ostringstream oss;
		output(oss);
		return oss.str();
	}
	friend ostream& operator<<(ostream& os, const grouplist& gs) {
		return os<<gs.hash;
	}
	inline size_t len() const {
		return groups.size();
	}
	bool operator<(const grouplist& r) const {
		const grouplist& me=*this;
		return me.len()!=r.len() ? me.len()<r.len() : me.hash<r.hash;
	}
	bool operator<=(const grouplist& r) const {
		const grouplist& me=*this;
		return me.len()!=r.len() ? me.len()<r.len() : me.hash<=r.hash;
	}
	bool operator>(const grouplist& r) const {
		const grouplist& me=*this;
		return me.len()!=r.len() ? me.len()>r.len() : me.hash>r.hash;
	}
	bool operator>=(const grouplist& r) const {
		const grouplist& me=*this;
		return me.len()!=r.len() ? me.len()>r.len() : me.hash>=r.hash;
	}
	bool operator==(const grouplist& r) const {
		const grouplist& me=*this;
		return me.len()!=r.len() ? false : me.hash==r.hash;
	}
	bool operator!=(const grouplist& r) const {
		const grouplist& me=*this;
		return me.len()!=r.len() ? true : me.hash!=r.hash;
	}
	group& operator[](int i) {
		size_t l=len();
		dbgcode (
			assert(l!=0);
		)
		if (i<0 || i>l) i=i % (int)l;
		if (i<0) i+=l;
		return groups[i];
	}
	grouplist slice(int i, int j) {
		size_t l=len();
		dbgcode (
			assert(l!=0);
		)
		if (i<0 || i>l) i=i % (int)l;
		if (i<0) i+=l;
		if (j<0 || j>l) j=j % (int)l;
		if (j<0) j+=l;
		return grouplist(groupvector(groups.begin()+i, groups.begin()+j));
	}
	void insert(int i, const group& g) {
		dbgcode (
			size_t l=len();
			assert(i>=0 && i<=l);
		)
		groups.insert(groups.begin()+i, g);
		update_hash();
	}
	void prepend(const group& g) {
		insert(0, g);
	}
	void append(const group& g) {
		insert(len(), g);
	}
	void del(int i) {
		size_t l=len();
		if (l==0) return;
		if (i<0 || i>=l) i=i % (int)l;
		if (i<0) i+=l;
		groups.erase(groups.begin()+i);
	}
private:
	ostream& output(ostream& os) {
		if (groups.size()!=0) {
			os<<groups[0];
			size_t l=groups.size();
			for (int i=1; i<l; ++i) os<<groups[i];//os<<'_'<<groups[i];
		}
		return os;
	}
};

int B(group& g) {
	dbgcode (
		assert(g.cnt>=1);
	)
	return g.cnt==1?2:1;
}

int B2(group& g) {
	dbgcode (
		assert(g.cnt>=2);
	)
	return g.cnt==2?1:0;
}

typedef map<grouplist, int> maptype;

maptype mf;
maptype mg;

inline bool key_in_map(grouplist& gs, maptype& m) {
	return m.count(gs)==1;
}

inline bool key_not_in_map(grouplist& gs, maptype& m) {
	return m.count(gs)==0;
}

int F(grouplist& gs, int layer=0) {
	// print(layer);
	// print(gs);
	// print(endl);
	if (key_in_map(gs, mf)) {
		return mf[gs];
	}
	size_t l=gs.len();
	int r=0;
	if (l==1) {
		group& C=gs[0];
		r=B(C);
	} else if (l==2) {
		group& L=gs[0];
		group& R=gs[1];
		r=B(L)+B(R);
	} else {
		group& L=gs[0];
		group& R=gs[-1];
		if (L.color==R.color) {
			grouplist C=gs.slice(1, -1);
			group tmp(L.color, L.cnt+R.cnt);
			r=B2(tmp)+F(C);
		} else {
			grouplist LC=gs.slice(0, -1);
			grouplist CR=gs.slice(1, l);
			int x=B(L)+F(CR, layer+1);
			int y=F(LC, layer+1)+B(R);
			r=min(x, y);
		}
	}
	// if (key_not_in_map(gs, mf)) 
	// print("result "<<gs<<"="<<r);
	return mf[gs]=r;
}

int G(grouplist& gs, int layer=0) {
	// print(layer);
	// print(gs);
	// print(endl);
	
	if (key_in_map(gs, mg)) {
		return mg[gs];
	}
	size_t l=gs.len();
	int r=0;
	if (l==0) {
		r=0;
	} else if (l<=3) {
		r=F(gs);
	} else {
		//test for any same color
		vector<char> colors;
		colors.reserve((int)l);
		for (int i=0; i<l; ++i) {
			colors.push_back(gs[i].color);
			r+=B(gs[i]);
		}
		sort(colors.begin(), colors.end());
		bool same=false;
		for (int i=0; i<l-1; ++i) {
			if (colors[i]==colors[i+1]) {
				same=true;
				break;
			}
		}
		if (!same) {
			mf[gs]=r;
		} else {
			//test for symmetry
			bool symmetry=false;
			if (l%2==1) {
				int c=0, m=l/2;
				while(c<m && gs[c].color==gs[l-1-c].color) ++c;
				if (c==m) symmetry=true;
			}
			if (symmetry) {
				r=F(gs);
				mf[gs]=r;
			} else {
				r=F(gs);
				size_t n=l;
				for (int radius=1; radius<n; ++radius) {
					for (int p=1; p<n-radius+2; ++p) {
						grouplist C=gs.slice(p-1, p-1+radius);
						if (radius>1 && radius%2==1  && C[0].color!=C[-1].color) continue;
						grouplist L=gs.slice(0, p-1);
						grouplist R=gs.slice(p-1+radius, l);
						
						// print(gs);
						// print(L);
						// print(C);
						// print(R);
						
						while (L.len()>0 && R.len()>0 && L[-1].color==R[0].color && L[-1].cnt+R[0].cnt>=3) {
							L.del(-1);
							R.del(0);
						}
						if (L.len()>0 && R.len()>0 && L[-1].color==R[0].color && L[-1].cnt+R[0].cnt<3) {
							group& g=R[0];
							group tmp=group(g.color, g.cnt);
							tmp.cnt+=L[-1].cnt;
							L.del(-1);
							R.del(0);
							L.append(tmp);
						}
						
						
						grouplist U=L+R;
						
						// print(L);
						// print(R);
						// print(U);
						// print(endl);
						
						int x=F(C)+G(U, layer+1);
						//
						if (x<r) r=x;
					}
				}
		
			}
		
		}
	}
	return mg[gs]=r;
}

void solve(string& s) {
	// print(s);
	grouplist gs;
	size_t l=s.size();
	group g;
	for (int i=0; i<l; ++i) {
		char c=s[i];
		if (g.cnt==0) {
			g.color=c;
			g.cnt=1;
		}
		if (i+1<l) {
			if (s[i+1]==c) {
				g.cnt+=1;
			} else {
				gs.append(g);
				g=group();
			}
		} else {
			gs.append(g);
		}
	}
	// print(gs);
	int r=F(gs);
	if (r>2) r=G(gs);
	cout<<r<<endl;
}


int main () {
	int n;
	while(in>>n) {
		string s;
		in>>s;
		solve(s);
		dbglog("ok");
	}
	
	// print(endl);
	// for (maptype::iterator it=mf.begin(); it!=mf.end(); ++it) {
		// cout<<it->first<<"\t"<<it->second<<endl;
	// }
	// print(endl);
	// for (maptype::iterator it=mg.begin(); it!=mg.end(); ++it) {
		// cout<< it->first <<"\t"<< it->second <<endl;
	// }
	
	return 0;
}

